### PR TITLE
poc: avro schema and records viewer

### PR DESCRIPTION
**Pull request by jsomsanith**
**What is the problem this PR is trying to solve?**
There is no object viewer that fits the avro schema and avro records format.

**What is the chosen solution to this problem?**
Make a generic object viewer.
Make a Schema viewer, based on generic one with customisation.
Make a Records viewer, based on generic one with customisation.
Use the same renderers as datagrid.
Add storybook to wrap up everything.

Storybook : http://tlnd-jso.surge.sh/?selectedKind=AvroViewer&selectedStory=Avro%20viewer
(on page load, the records height are not right because of style that is loaded after render. So the mesurement is not ok. To fix that, select another story in the left menu and select back the avro viewer story)

To go over this PoC, you can start with the [PoC guideline](https://github.com/Talend/ui/blob/ea1e15f35608d75301282b94d027dad335eb3874/packages/components/src/ObjectViewer/AVRO.README.MD)

**Please check if the PR fulfills these requirements**
- [ ] The PR commit message follows our [guidelines](https://github.com/talend/tools/blob/master/tools-root-github/CONTRIBUTING.md)
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)
- [ ] Related design / discussions / pages (not in jira), if any, are all linked or available in the PR

**[ ] This PR introduces a breaking change**

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -18,6 +18,7 @@
     "watch": "npm-watch",
     "publish": "lerna publish --force-publish=*",
     "build-storybook": "npm run test:demo && export ACTION='test:demo' && export TRAVIS_BRANCH='master' && export TRAVIS_BUILD_DIR=`pwd` && .travis/after_success_static.sh && .travis/after_success_demo.sh && .travis/after_success_coverage.sh",
+    "publish-jso": "cd packages/components/storybook-static && surge -d tlnd-jso.surge.sh",
```
Comment (romainseb) [inline package.json]: you forgot this :)

File: output/components.eslint.txt [hunk 2]
```diff
@@ -15,9 +15,24 @@ The react/require-extension rule is deprecated. Please use the import/extensions
 /home/travis/build/Talend/ui/packages/components/src/JSONSchemaRenderer/JSONSchemaRenderer.component.js
   76:56  error  'properties' is defined but never used  no-unused-vars
 
+/home/travis/build/Talend/ui/packages/components/src/ObjectViewer/AvroRenderer/index.js
+  22:1  error  Prefer default export  import/prefer-default-export
```
Comment (romainseb) [inline output/components.eslint.txt]: Could you fix some added lint ?

File: packages/components/src/ObjectViewer/Model/genericViewer.configuration.js [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+ * Here we want to display the model
+ * - item = array : the model is an array of field. This condition for root node.
+ * 					we want to display each root property as a JSON object property
+ * - item.fields : 	it means that it is a record.
+ * 					its children fields should be displayed as a JSON object property
+ * - item.items : 	it means that it is an array.
+ * 				 	but we want to display the array objects fields as a JSON object property
+ */
+function getDataType(item) {
+	return Array.isArray(item) || item.fields || item.items ? 'object' : null;
```
Comment (talend-glorieux) [inline packages/components/src/ObjectViewer/Model/genericViewer.configuration.js]: Simplify.

File: packages/components/src/ObjectViewer/Model/genericViewer.configuration.js [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+			0: 5,
+			1: 65,
+			'-1': 30,
+		},
+	},
+ * We display the DQ type if present and fallback to the simple type.
+ */
+function getDisplayKey({ value }) {
+	const type = value.type && (value.type.dqType || value.type.type);
+	return [value.doc, type && <span className={theme.type}>({type})</span>];
```
Comment (talend-glorieux) [inline packages/components/src/ObjectViewer/Model/genericViewer.configuration.js]: Should be simplified.

File: packages/components/src/ObjectViewer/Model/genericViewer.configuration.js [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+	let fields;
+	if (Array.isArray(item)) {
+		fields = item;
+	} else {
+		fields = item.fields || (item.items && item.items.fields);
+	}
+	if (fields) {
+		return fields.map(field => ({ dataKey: field.name, value: field }));
+	}
+	return null;
```
Comment (talend-glorieux) [inline packages/components/src/ObjectViewer/Model/genericViewer.configuration.js]: Start with the error case.

File: packages/components/src/ObjectViewer/ObjectViewer.component.js [hunk 6]
```diff
@@ -1,15 +1,21 @@
...
+import GenericViewer from './Generic';
 import JSONLike from './JSONLike';
 import List from './List';
+import Model from './Model';
+import Records from './Records';
+import Table from './Table';
 
 export const DISPLAY_MODES = {
 	FLAT: 'flat',
+	GENERIC: 'generic',
```
Comment (talend-glorieux) [inline packages/components/src/ObjectViewer/ObjectViewer.component.js]: 'Generic' is too generic a name.

File: packages/components/src/ObjectViewer/ObjectViewer.component.js [hunk 7]
```diff
@@ -1,15 +1,21 @@
...
+	LIST: 'list',
+	MODEL: 'model',
+	RECORDS: 'records',
 	TABLE: 'table',
 	TREE: 'tree',
```
Comment (talend-glorieux) [inline packages/components/src/ObjectViewer/ObjectViewer.component.js]: What's the difference between the `records` and `tree` view?

File: packages/components/src/ObjectViewer/Records/RecordRenderer.component.js [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+	const cache = parent.props.cache;
+	const highlighted = parent.props.highlighted;
+	const onToggle = parent.props.onRowItemToggle;
+	const opened = parent.props.opened[index];
+
+	const schema = parent.props.schema;
+	const datum = parent.props.data[index];
+
+	const getComponent = parent.props.getComponent;
+	const avroRenderersIds = parent.props.avroRenderersIds;
```
Comment (talend-glorieux) [inline packages/components/src/ObjectViewer/Records/RecordRenderer.component.js]: Useless constants.

File: packages/components/src/ObjectViewer/Records/RecordViewer.component.js [hunk 9]
```diff
@@ -0,0 +1,22 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import genericViewerConfiguration from './genericViewer.configuration';
+import GenericViewer from '../Generic';
+
+import theme from './RecordViewer.scss';
+
+export default function Record(props) {
+	const data = { schema: props.schema, data: props.data };
```
Comment (talend-glorieux) [inline packages/components/src/ObjectViewer/Records/RecordViewer.component.js]: Useless constant. `data` is only used once.

File: packages/components/src/ObjectViewer/Records/genericViewer.configuration.js [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+	return Array.isArray(avroSample.data.value) ? 'array' : typeof avroSample.data.value;
+}
+
+/**
+ * The value is an avro entry that contains { schema, data }.
+ * The data is an avro data, containing a quality indicator.
+ * We adapt this indicator to a quality constant.
+ */
+function getQuality(props) {
+	const avroSample = props.value;
```
Comment (talend-glorieux) [inline packages/components/src/ObjectViewer/Records/genericViewer.configuration.js]: Useless constant. 
You should use something like lodash's `getIn`.

File: packages/components/src/ObjectViewer/Records/genericViewer.configuration.js [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+	return avroSample.data.quality;
+}
+
+/**
+ * The value is an avro entry that contains { schema, data }.
+ * The data is an avro data, containing the real value.
+ */
+function getDisplayValue(props) {
+	const avroSample = props.value;
```
Comment (talend-glorieux) [inline packages/components/src/ObjectViewer/Records/genericViewer.configuration.js]: Useless constant.

File: packages/components/src/ObjectViewer/Records/genericViewer.configuration.js [hunk 12]
```diff
@@ -0,0 +1,128 @@
...
+	let className;
+	if (type === 'object') {
+		name = 'talend-plus-circle'; // TODO we don't have a talend-minus-circle
+		className = theme.icon;
+	} else {
+		name = isOpened ? 'talend-caret-down' : 'talend-chevron-left';
+		transform = isOpened ? null : 'rotate-180';
+	}
+
+	return { className, name, transform };
```
Comment (talend-glorieux) [inline packages/components/src/ObjectViewer/Records/genericViewer.configuration.js]: Please refrain from using variables in favor of clear return cases.

File: packages/components/stories/AvroViewer.js [hunk 13]
```diff
@@ -0,0 +1,605 @@
...
+		type: {
+			type: 'string',
+			dqType: 'Recipe',
+			dqTypeKey: 'RECIPE',
+		},
+		'@talend-quality@': {
+			0: 5,
+			1: 65,
+			'-1': 30,
+		},
```
Comment (talend-glorieux) [inline packages/components/stories/AvroViewer.js]: Wait a minute. :thinking: 
This isn't an Avro Schema. 
It looks like a weird kind of wrapped, hand-made JSON.

File: packages/components/stories/AvroViewer.js [hunk 14]
```diff
@@ -0,0 +1,605 @@
...
+	],
+	valid: [
+		{ label: 'Filter valid values', onClick: recordFilterAction },
+		{ label: 'Remove valid values', onClick: recordRemovalAction },
+	],
+};
+
+const modelItemMenu = [
+	{ label: 'LOL action', onClick: action('onLolClick') },
+	{ label: 'MDR action', onClick: action('onMdrClick') },
```
Comment (talend-glorieux) [inline packages/components/stories/AvroViewer.js]: Please rename.

File: packages/containers/src/AvroViewer/AvroViewer.container.js [hunk 15]
```diff
@@ -0,0 +1,75 @@
...
+	constructor(props) {
+		super(props);
+		this.onSelect = this.onSelect.bind(this);
+	}
+
+	onSelect(event, jsonpath) {
+		const state = this.props.state || DEFAULT_STATE;
+		const adaptedJsonPath = jsonpath
+			.replace(/[-[{}()*+?.,\\^$|#\s]/g, '\\$&')
+			.replace(/\[]/g, '[[0-9]+]');
```
Comment (talend-glorieux) [inline packages/containers/src/AvroViewer/AvroViewer.container.js]: Why do you need to convert the JSONPath? Can't we use AvroPath instead?

File: packages/components/src/ObjectViewer/AvroRenderer/DefaultAvroRenderer.component.js [hunk 16]
```diff
@@ -0,0 +1,38 @@
...
+
+const defaultRenderers = {
+	date: DefaultDateRenderer,
+};
+
+function getDefaultRenderer(type) {
+	return defaultRenderers[type] || DefaultRenderer;
+}
+
+export default function DefaultAvroRenderer({ avroRenderersIds, colDef, data, getComponent }) {
```
[the remark this answers is not in the document]
Comment (jsomsanith) [reply packages/components/src/ObjectViewer/AvroRenderer/DefaultAvroRenderer.component.js]: Yes should be renamed, this was done to reuse the renderers of `datagrid` module.
It's a fieldDefinition, that contains the avro schema of this field.

File: packages/components/src/ObjectViewer/AvroRenderer/index.js [hunk 17]
```diff
@@ -0,0 +1,27 @@
+import React from 'react';
+import DefaultAvroRenderer from './DefaultAvroRenderer.component';
+import Inject from '../../Inject';
+
+// Each time we call injectedCellRenderer, it produces a new component.
+// React consider it as a change, and change the dom at each render :/
+// We can't memoize the because we don't have unique id per instance
+// TODO need to find a solution
```
[the remark this answers is not in the document]
Comment (jsomsanith) [reply packages/components/src/ObjectViewer/AvroRenderer/index.js]: Agree

File: packages/components/src/ObjectViewer/Generic/GenericViewer.component.js [hunk 18]
```diff
@@ -0,0 +1,323 @@
...
+		isOpened,
+		jsonpath,
+		dataKey,
+		onClick,
+		onToggle,
+		style,
+		type,
+	} = props;
+
+	const content = [<span key={'datakey'}>{getDisplayKey(props)}</span>];
```
[the remark this answers is not in the document]
Comment (jsomsanith) [reply packages/components/src/ObjectViewer/Generic/GenericViewer.component.js]: Come on, we all know that const on objects are not constant.
Right ? :D

File: packages/components/src/ObjectViewer/Generic/genericViewer.configuration.js [hunk 19]
```diff
@@ -0,0 +1,45 @@
...
+
+export function defaultGetJSONPath({ dataKey, parent }) {
+	return getJSONPathUtils(dataKey, parent.jsonpath, parent.type);
+}
+
+export function defaultGetQuality() {
+	return null;
+}
+
+export function defaultGetIcon({ isOpened }) {
```
[the remark this answers is not in the document]
Comment (jsomsanith) [reply packages/components/src/ObjectViewer/Generic/genericViewer.configuration.js]: It is for usage readability.
Those functions are the default implementations of customisation entry points.
So in GenericViewer, we import them and use them as default (if no custom implementations are passed via props). So in the code, reading `defaultMyFn` vs `props.myFn` if clearer than `myFn` vs `props.myFn`

File: packages/components/src/ObjectViewer/Generic/genericViewer.configuration.js [hunk 20]
```diff
@@ -0,0 +1,45 @@
...
+	}
+	return value;
+}
+
+export function defaultGetJSONPath({ dataKey, parent }) {
+	return getJSONPathUtils(dataKey, parent.jsonpath, parent.type);
+}
+
+export function defaultGetQuality() {
+	return null;
```
[the remark this answers is not in the document]
Comment (jsomsanith) [reply packages/components/src/ObjectViewer/Generic/genericViewer.configuration.js]: 👍 this goes with your previous comment on extracting quality values into constants.

File: packages/components/src/ObjectViewer/Model/ModelItemMenu.component.js [hunk 21]
```diff
@@ -0,0 +1,46 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import { Action } from '../../Actions';
+
+export default class ModelItemMenu extends React.Component {
```
[the remark this answers is not in the document]
Comment (jsomsanith) [reply packages/components/src/ObjectViewer/Model/ModelItemMenu.component.js]: It is a component that renders a menu for each node in the model viewer.
Is it the name that you find confusing ?

File: packages/components/src/ObjectViewer/ObjectViewer.component.js (hunk 7 above)
Comment (jsomsanith) [reply packages/components/src/ObjectViewer/ObjectViewer.component.js]: Records display a list of record and each record is displayed with a viewer based on the generic one.
Generic includes quality indicator, that is used in model viewer too.
The idea behind the genericViewer is that it should be the common part of all these type of viewers, and allow each specific viewer to be based on it with some customisation. In a perfect world, tree and object would be refactored to use GenericViewer

File: packages/containers/src/AvroViewer/AvroViewer.container.js (hunk 15 above)
Comment (jsomsanith) [reply packages/containers/src/AvroViewer/AvroViewer.container.js]: As you said, the format is not really the avro standard but a modified version.
Plus, how we render the nodes is recursive, we don't have the entire record everytime, we don't get the value from root, but from the sub-record we have.

**Pull request closed without merging**